docs/md/guides: database per tenant (#2990)
ariga/atlas: doc/md/guides/database-per-tenant/04-cloud.mdx

databases. To do this, let's make a small change to our `prod` env in `atlas.hcl`:

```hcl
env "prod" {
  for_each = toset(local.tenant)
  url = "sqlite://${each.value}.db"
  migration {
//highlight-next-line
    dir = "atlas://db-per-tenant"
  }
}
```

Now, we can deploy the migrations to our target databases by running:

```shell
atlas migrate apply --env prod
```

Atlas will read the most recent version of our migration directory from the schema registry, apply the migrations
to each target database, report the results to Atlas Cloud, and print the results:

```
No migration files to execute
No migration files to execute
https://rotemtam85.atlasgo.cloud/deployments/sets/94489280593
```

In this case, we see that there were no new migrations to apply to the target databases. Let's show how
this flow works when there is work to be done in the next section.

### Another migration

Let's plan another migration to our project. Create a new migration file by running:

```shell
atlas migrate new --edit seed_users
```

In the editor, add the following SQL statements:

```sql
INSERT INTO users (id, name) VALUES (1, "a8m");
INSERT INTO users (id, name) VALUES (2, "rotemtam");
```

> Save the file and exit the editor. Let's now push the new migration to the Schema Registry:

Reviewer: ```suggestion
Save the file and exit the editor. Let's push the new migration to the Schema Registry:
```

You can see the status of each deployment, the target databases, and the results of each migration.

### Database Status

![](https://atlasgo.io/uploads/db-per-tenant/databases-screen.png)

To view the status of the different databases in your project, navigate to the "Databases" tab in the Atlas Cloud
UI. Here, you can see the status of each database, the most recent migration applied, and the results of the
migration.

Databases can be in one of three states:
* Synced - The database is up-to-date with the most recent migration.
* Pending - The database is waiting for a new migration to be applied.
* Error - An error occurred while applying the migration.

### Troubleshooting

If an error occurs during a migration, having a centralized view of all your deployments can help you quickly
identify the issue and take corrective action. You can view the error message, the target database, and the
migration that caused the error.

Suppose we run a deployment that fails during the schema migration phase, we can easily locate the error in the
Atlas Cloud UI by navigating to the "Migrations" tab:

![](https://atlasgo.io/uploads/db-per-tenant/migrations-screen.png)

> We easily locate the failed deployment and drill down to diagnose the issue:

Reviewer: ```suggestion
We quickly find the failed deployment and drill down to diagnose the issue:
```

![](https://atlasgo.io/uploads/db-per-tenant/deployment-set-error.png)

From the logs, we see that 3 out of 4 migrations passed without action, but the last one failed. We see that it failed
on `tenant_4.db` with the error message:

```
Error: sql/migrate: executing statement "INSERT INTO users (id, name) VALUES (1, \"a8m\");" from version "20240721111345": UNIQUE constraint failed: users.id
```
We can further drill down into the specific database target migration:

![](https://atlasgo.io/uploads/db-per-tenant/deployment-error.png)

> We now clearly see the issue, our data migration failed due to a unique constraint violation. We can now take corrective

Reviewer: ```suggestion
We now clearly see the issue, our data migration failed due to a unique constraint violation. Now, we can take corrective
```

action to fix the issue and reapply the migration - usually by fixing the problematic data in our target database.

## Conclusion

In this section, we demonstrated how to use the Atlas Cloud Control Plane to manage migrations across multiple
target databases. We showed how to push our project to the Atlas Cloud Schema Registry, deploy migrations to
target databases, and gain visibility into the status of our deployments.

While it is possible to manage migrations using the Atlas CLI, the Atlas Cloud Control Plane provides a centralized
view of all your deployments, making it easier to manage and troubleshoot issues across multiple databases.

> 

Reviewer: ```suggestion
```
